src/blocks/sound/pipewire.rs

```diff
@@ -0,0 +1,645 @@
+                            }
+                        )
+                        .expect("Unable to add object to pod");
+                        let pod = Pod::from_bytes(&pod_data).expect("Unable to construct pod");
+                        proxy.set_param(ParamType::Props, 0, pod);
+                    }
+                }
+            }
+            Mute(id, mute) => {
+                if let Some(node) = client.data.lock().unwrap().nodes.get(&id) {
+                    if let Some(proxy) = proxies.borrow_mut().proxies_t.get(&node.proxy_id) {
+                        let mut pod_data = Vec::new();
+                        let mut pod_builder = PodBuilder::new(&mut pod_data);
+                        builder_add!(&mut pod_builder,
+                            Object(
+                                SpaTypes::ObjectParamProps.as_raw(),
+                                ParamType::Props.as_raw()
+                            ) {
+                                SPA_PROP_mute => Bool(mute),
+                                SPA_PROP_softMute => Bool(mute),
+                            }
+                        )
+                        .expect("Unable to add object to pod");
+                        let pod = Pod::from_bytes(&pod_data).expect("Unable to construct pod");
+                        proxy.set_param(ParamType::Props, 0, pod);
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct Client {
+    event_listeners: Mutex<Vec<Sender<EventKind>>>,
+    command_listeners: Mutex<Vec<Receiver<CommandKind>>>,
+    pub ready: Mutex<bool>,
+    pub data: Mutex<Data>,
+}
+
+impl Client {
+    fn new() -> Result<Client> {
+        thread::Builder::new()
+            .name("privacy_pipewire".to_string())
+            .spawn(Client::main_loop_thread)
+            .error("failed to spawn a thread")?;
+
+        Ok(Client::default())
+    }
+
+    fn spawn_idle_listener(
+        loop_ref: &LoopRef,
+        proxies_weak: Weak<RefCell<NodeProxies>>,
+    ) -> IdleSource {
+        let client = CLIENT.as_ref().error("Could not get client").unwrap();
+        loop_ref.add_idle(true, move || {
+            if let Some(proxies) = proxies_weak.upgrade() {
+                let mut commands = client.command_listeners.lock().unwrap();
+                let mut i = 0;
+                while i < commands.len() {
+                    let recv_result = commands[i].try_recv();
+                    if matches!(recv_result, Err(TryRecvError::Disconnected)) {
+                        commands.remove(i);
+                        continue;
+                    }
+                    i += 1;
+                    let Ok(command) = recv_result else {
+                        continue;
+                    };
+                    return command.execute(client, proxies);
+                }
+            }
+            std::thread::sleep(Duration::from_millis(1));
```

Comment:
Why is this sleep() needed?

Reply:
Since the loop runs repeatedly, the sleep prevents pegging the cpu